packages/react-compose/src/types.ts

```diff
@@ -13,23 +13,27 @@ export interface ShorthandConfig<TProps> {
 
 export type PropsOfElement<
   // tslint:disable-next-line:no-any
-  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
-> = JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
+  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any> | ComponentWithAs
+> = E extends { __PRIVATE_PROPS: any }
+  ? E['__PRIVATE_PROPS']
+  : JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
```

Comment:
It's a hack for `ComponentWithAs` as otherwise it will be resolved to:

```
(props: Record<any, any> & { as?: React.ElementType } & SomeProps) => JSX.Element
```

As TS struggles to resolve `TExtendedElementType`.

***

The idea of `__PRIVATE_PROPS` is based on the assumption that it's possible to use `as` only once:

```tsx
<Button /> => // button 
<Button as='div' /> => // div
<Button as={Link} /> => // a
// ??? No way to pass `as` for `Link` 👍 
```

In this case we can assign `__PRIVATE_PROPS` to a default value of `TElementType` to simplify the resolutions.